core/src/main/java/org/infinispan/atomic/impl/TransactionHelper.java

```diff
@@ -0,0 +1,84 @@
+package org.infinispan.atomic.impl;
+
+import java.util.function.Supplier;
+
+import javax.transaction.SystemException;
+import javax.transaction.Transaction;
+import javax.transaction.TransactionManager;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.atomic.FineGrainedAtomicMap;
+import org.infinispan.batch.BatchContainer;
+import org.infinispan.commons.CacheException;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+class TransactionHelper {
+   private static final Log log = LogFactory.getLog(FineGrainedAtomicMap.class);
+   private static final boolean trace = log.isTraceEnabled();
+
+   private final TransactionManager transactionManager;
+   private final BatchContainer batchContainer;
+   private final boolean autoCommit;
+
+   public TransactionHelper(AdvancedCache<?, ?> cache) {
+      this.transactionManager = cache.getTransactionManager();
+      if (transactionManager == null) {
+         throw log.atomicFineGrainedNeedsTransactions();
+      }
+      this.batchContainer = cache.getCacheConfiguration().invocationBatching().enabled() ? cache.getBatchContainer() : null;
+      this.autoCommit = cache.getCacheConfiguration().transaction().autoCommit();
+   }
+
+   private Transaction getOngoingTransaction() {
+      try {
+         Transaction transaction = null;
+         if (transactionManager != null) {
+            transaction = transactionManager.getTransaction();
+            if (transaction == null && batchContainer != null) {
+               transaction = batchContainer.getBatchTransaction();
+            }
+         }
+         return transaction;
+      } catch (SystemException e) {
+         throw new CacheException("Unable to get transaction", e);
+      }
+   }
+
+   <T> T run(Supplier<T> op) {
```

Comment:
Then we should fix `CacheImpl.executeCommandAndCommitIfNeeded` and some code in functional maps as well... this is basically inlined ^.